internals/overlord/logstate/gatherer.go

```diff
@@ -154,21 +173,26 @@ func (g *logGatherer) PlanChanged(pl *plan.Plan, buffers map[string]*servicelog.
 			continue
 		}
 
-		buffer, bufferExists := buffers[service.Name]
-		if !bufferExists {
+		svcData, svcStarted := serviceData[service.Name]
+		if !svcStarted {
 			// We don't yet have a reference to the service's ring buffer
 			// Need to wait until ServiceStarted
 			continue
 		}
 
-		g.pullers.Add(service.Name, buffer, g.entryCh)
+		g.pullers.Add(service.Name, svcData.Buffer, g.entryCh)
+		// update labels
+		g.setLabels <- svcWithLabels{service.Name, labels[service.Name]}
 	}
 }
 
 // ServiceStarted is called by the LogManager on the start of a service which
 // logs to this gatherer's target.
-func (g *logGatherer) ServiceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+func (g *logGatherer) ServiceStarted(service *plan.Service, buffer *servicelog.RingBuffer, labels map[string]string) {
```

Comment:
Update the comment above to say that ServiceStarted expects labels to never change after it is passed in. Otherwise we need to copy it before we send it on the channel.